rust/src/http2/http2.rs

```diff
@@ -0,0 +1,627 @@
+        let mut tx = HTTP2Transaction::new();
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        return tx;
+    }
+
+    fn parse_ts(&mut self, mut input: &[u8]) -> AppLayerResult {
+        //first consume frame bytes
+        let il = input.len();
+        if self.request_frame_size > 0 {
+            let ilen = input.len() as u32;
+            if self.request_frame_size >= ilen {
+                self.request_frame_size -= ilen;
+                return AppLayerResult::ok();
+            } else {
+                let start = self.request_frame_size as usize;
+                input = &input[start..];
+                self.request_frame_size = 0;
+            }
+        }
+
+        //then parse all we can
+        while input.len() > 0 {
+            match parser::http2_parse_frame_header(input) {
+                Ok((rem, head)) => {
+                    //TODO handle transactions the right way
+                    let mut tx = self.new_tx();
+                    tx.ftype = Some(head.ftype);
+                    match head.ftype {
+                        parser::HTTP2FrameType::GOAWAY => {
+                            match parser::http2_parse_frame_goaway(rem) {
+                                Ok((_, goaway)) => {
+                                    tx.type_data = Some(HTTP2FrameTypeData::GOAWAY(goaway));
+                                }
+                                Err(nom::Err::Incomplete(_)) => {
+                                    return AppLayerResult::incomplete(
+                                        (il - input.len()) as u32,
+                                        4 as u32,
+                                    );
+                                }
+                                Err(_) => {
+                                    self.set_event(HTTP2Event::InvalidFrameData);
+                                }
+                            }
+                        }
+                        //TODO parse deeper based on other frame type
+                        _ => {}
+                    }
+                    self.transactions.push(tx);
+
+                    let hl = head.length as usize;
+                    if rem.len() < hl {
+                        let rl = rem.len() as u32;
+                        self.request_frame_size = head.length - rl;
+                        return AppLayerResult::ok();
+                    } else {
+                        input = &rem[hl..];
+                    }
+                }
+                Err(nom::Err::Incomplete(_)) => {
+                    //we may have consumed data from previous records
+                    if input.len() < HTTP2_FRAME_HEADER_LEN {
+                        return AppLayerResult::incomplete(
+                            (il - input.len()) as u32,
+                            HTTP2_FRAME_HEADER_LEN as u32,
+                        );
+                    } else {
+                        //TODO BUG_ON ?
```

Comment:
Is there something equivalent to `BUG_ON` in Rust ? @jasonish

Reply:
`panic!("msg");`

Reply:
ok thanks